### PR TITLE
Disable shared bundles for development mode

### PR DESCRIPTION
**Pull request by AGawrys**
# ↪️ Pull Request

This pull request is for disabling shared bundles in development. This change reduces build time for development. Tests testing shared bundle behavior have been modified to run in production mode and in some cases with scope hoisting and minify off.

## 🚨 Test instructions

Added a test case that confirms duplicated bundles in development. 

Run yarn test and confirm all test cases pass

### CODE REVIEW

File: packages/core/integration-tests/test/html.js [hunk 1]
```diff
@@ -1058,7 +1058,7 @@ describe('html', function() {
         __dirname,
         '/integration/html-js-shared-dynamic-nested/index.html',
       ),
-      {production: true, scopeHoist: true},
+      {mode: 'production', production: true, scopeHoist: true},
```
Comment (mischnic) [inline packages/core/integration-tests/test/html.js:1061]: Can you remove these `production: true` properties? These are leftovers from Parcel 1. And looks like these tests actually once ran in production mode 😄

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.48s | -156.00ms |
| Cached | 910.00ms | +14.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 43.00ms | -23.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 78.00ms | +36.00ms ⚠️ |
| dist/legacy/index.6fa78f1c.js | 1.07kb | +0.00b | 1.44s | -125.00ms 🚀 |
| dist/legacy/index.afdbda6b.css | 77.00b | +0.00b | 1.27s | -163.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 71.00ms | -23.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 71.00ms | +7.00ms ⚠️ |
| dist/modern/index.df30fe40.js | 1.07kb | +0.00b | 93.00ms | -7.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 28.06s | -125.00ms |
| Cached | 15.43s | +491.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/NotFound.141e0b29.js | 532.00b | +0.00b | 342.00ms | +26.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 53.00ms | +11.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 479.93kb | +0.00b | 81.00ms | +16.00ms ⚠️ |
| dist/PermalinkedComment.a5120a24.js | 4.22kb | +0.00b | 81.00ms | +15.00ms ⚠️ |
| dist/UserProfile.95e14af2.js | 1.70kb | +0.00b | 81.00ms | +15.00ms ⚠️ |
| dist/NotFound.141e0b29.js | 532.00b | +0.00b | 61.00ms | -4.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.55m | +435.00ms |
| Cached | 3.42s | -81.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/card.3fd8a427.js | 5.71kb | +0.00b | 13.73s | +1.23s ⚠️ |
| dist/dropzone.dcab7b1b.js | 3.44kb | +0.00b | 26.64s | +1.57s ⚠️ |
| dist/Modal.d8614cea.js | 3.15kb | +0.00b | 1.28s | +140.00ms ⚠️ |
| dist/16.4fc0f868.js | 2.49kb | +0.00b | 1.28s | +140.00ms ⚠️ |
| dist/card.bfb3f5f8.js | 2.15kb | +0.00b | 24.01s | +8.27s ⚠️ |
| dist/code.17026d12.js | 1.61kb | +0.00b | 5.08s | -1.89s 🚀 |
| dist/16.55d1de3e.js | 1.45kb | +0.00b | 1.28s | +140.00ms ⚠️ |
| dist/16.341ad0e1.js | 1.34kb | +0.00b | 1.28s | +140.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.9f8e76f5.js | 1.30kb | +0.00b | 13.73s | +1.23s ⚠️ |
| dist/16.8015c4e6.js | 1.27kb | +0.00b | 1.28s | +140.00ms ⚠️ |
| dist/16.2e0d51ab.js | 1.18kb | +0.00b | 6.91s | +1.79s ⚠️ |
| dist/media-viewer-analytics-error-boundary.bc280d38.js | 995.00b | +0.00b | 24.01s | +8.27s ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.4a94163b.js | 2.31mb | +0.00b | 639.00ms | -54.00ms 🚀 |
| dist/popup.0686950e.js | 169.21kb | +0.00b | 772.00ms | -39.00ms 🚀 |
| dist/Modal.8e9ccd89.js | 38.79kb | +0.00b | 643.00ms | +81.00ms ⚠️ |
| dist/component.f31cba5f.js | 30.87kb | +0.00b | 589.00ms | +44.00ms ⚠️ |
| dist/component.c8d4c7aa.js | 22.40kb | +0.00b | 614.00ms | +52.00ms ⚠️ |
| dist/js.45b69177.js | 16.55kb | +0.00b | 613.00ms | +51.00ms ⚠️ |
| dist/component.aaadd915.js | 6.27kb | +0.00b | 614.00ms | +35.00ms ⚠️ |
| dist/media-viewer.b91b6ee1.js | 4.01kb | +0.00b | 772.00ms | -39.00ms 🚀 |
| dist/png-chunks-extract.417e691a.js | 3.55kb | +0.00b | 614.00ms | +52.00ms ⚠️ |
| dist/dropzone.dcab7b1b.js | 3.44kb | +0.00b | 772.00ms | -39.00ms 🚀 |
| dist/Modal.d8614cea.js | 3.15kb | +0.00b | 638.00ms | -55.00ms 🚀 |
| dist/16.4fc0f868.js | 2.49kb | +0.00b | 589.00ms | +45.00ms ⚠️ |
| dist/card.bfb3f5f8.js | 2.15kb | +0.00b | 772.00ms | -39.00ms 🚀 |
| dist/feedback.89b37c2d.js | 1.86kb | +0.00b | 727.00ms | -48.00ms 🚀 |
| dist/16.5678cfe5.js | 1.86kb | +0.00b | 670.00ms | +45.00ms ⚠️ |
| dist/16.fb8534c0.js | 1.79kb | +0.00b | 638.00ms | +59.00ms ⚠️ |
| dist/16.85df7a6e.js | 1.51kb | +0.00b | 670.00ms | +74.00ms ⚠️ |
| dist/16.34fd12e7.js | 1.46kb | +0.00b | 638.00ms | +59.00ms ⚠️ |
| dist/16.ff3f3afd.js | 1.46kb | +0.00b | 638.00ms | +43.00ms ⚠️ |
| dist/16.55d1de3e.js | 1.45kb | +0.00b | 589.00ms | +44.00ms ⚠️ |
| dist/16.9689f28f.js | 1.44kb | +0.00b | 656.00ms | -54.00ms 🚀 |
| dist/16.341ad0e1.js | 1.34kb | +0.00b | 589.00ms | +45.00ms ⚠️ |
| dist/Modal.80b784cd.js | 1.28kb | +0.00b | 613.00ms | +51.00ms ⚠️ |
| dist/16.8015c4e6.js | 1.27kb | +0.00b | 589.00ms | +45.00ms ⚠️ |
| dist/16.1e3b9f3d.js | 1.27kb | +0.00b | 670.00ms | +74.00ms ⚠️ |
| dist/16.30604232.js | 1.26kb | +0.00b | 638.00ms | +59.00ms ⚠️ |
| dist/16.a86c2eba.js | 1.26kb | +0.00b | 670.00ms | +75.00ms ⚠️ |
| dist/16.8e47d46b.js | 1.26kb | +0.00b | 670.00ms | +74.00ms ⚠️ |
| dist/16.867d534c.js | 1.26kb | +0.00b | 637.00ms | +58.00ms ⚠️ |
| dist/component.0a97dc28.js | 1.23kb | +0.00b | 638.00ms | -56.00ms 🚀 |
| dist/16.a43add3b.js | 1.23kb | +0.00b | 638.00ms | -55.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.bc280d38.js | 995.00b | +0.00b | 772.00ms | -39.00ms 🚀 |
| dist/index.html | 119.00b | +0.00b | 559.00ms | -54.00ms 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 19.22s | -368.00ms |
| Cached | 804.00ms | -7.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 580.68kb | +0.00b | 124.00ms | +9.00ms ⚠️ |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/285280736179651080)**

File: packages/core/integration-tests/test/html.js (hunk 1 above)
Comment (AGawrys) [reply packages/core/integration-tests/test/html.js:1061]: Yes! I wasn't sure if that should be a separate task, but I'm on it 🙇‍♀️  @mischnic

Comment (AGawrys) [reply packages/core/integration-tests/test/html.js]: I resolved this and updated the branch but I'm not sure why the checks aren't passing 🤧  @mischnic

File: packages/core/core/src/BundlerRunner.js [hunk 2]
```diff
@@ -125,12 +125,14 @@ export default class BundlerRunner {
     // $FlowFixMe
     await dumpGraphToGraphViz(internalBundleGraph._graph, 'after_bundle');
     try {
-      await bundler.optimize({
-        bundleGraph: mutableBundleGraph,
-        config: configResult?.config,
-        options: this.pluginOptions,
-        logger: new PluginLogger({origin: this.config.getBundlerName()}),
-      });
+      if (this.pluginOptions.mode === 'production') {
```
Comment (mischnic) [inline packages/core/core/src/BundlerRunner.js:128]: nit: I'd have put the if outside the try/catch.

File: packages/bundlers/default/src/DefaultBundler.js [hunk 3]
```diff
@@ -10,6 +10,7 @@ import type {
 import type {SchemaEntity} from '@parcel/utils';
 
 import invariant from 'assert';
+//
```
Comment (wbinnssmith) [inline packages/bundlers/default/src/DefaultBundler.js:13]: leftover from something?

File: packages/bundlers/default/src/DefaultBundler.js [hunk 4]
```diff
@@ -185,9 +186,7 @@ export default (new Bundler({
         bundleGraph.addEntryToBundle(asset, bundle);
       }
     }
-  },
-
-  optimize({bundleGraph, config}) {
+    // Optimizations
```
Comment (wbinnssmith) [inline packages/bundlers/default/src/DefaultBundler.js:189]: this is a bit confusing given the `optimize` method below 😅

File: packages/core/core/src/BundlerRunner.js [hunk 5]
```diff
@@ -127,17 +127,19 @@ export default class BundlerRunner {
...
-        bundleGraph: mutableBundleGraph,
-        config: configResult?.config,
-        options: this.pluginOptions,
-        logger: new PluginLogger({origin: this.config.getBundlerName()}),
-      });
-    } catch (e) {
-      throw new ThrowableDiagnostic({
-        diagnostic: errorToDiagnostic(e, this.config.getBundlerName()),
-      });
+    if (this.pluginOptions.mode === 'production') {
```
Comment (wbinnssmith) [inline packages/core/core/src/BundlerRunner.js:130]: the signal check and `after_optimize` graphviz dump can be moved in here as well

File: packages/packagers/js/src/prelude.js [hunk 6]
```diff
@@ -24,6 +24,8 @@
   var previousRequire =
     typeof globalObject[parcelRequireName] === 'function' &&
     globalObject[parcelRequireName];
+
+  cache = previousRequire.cache || {};
```
Comment (wbinnssmith) [inline packages/packagers/js/src/prelude.js:28]: can we avoid passing the cache in as an argument to the main prelude iife and just initialize it here in the case there's no previous require?

cc @devongovett

**Review by wbinnssmith — APPROVED** [review]

**Pull request merged**